### ISSUE TITLE
LargeMarginSoftmaxLoss is casting the W to float16 when using Mixed Precision

### ISSUE THREAD

**Issue by fernandocamargoti** [KevinMusgrave/pytorch-metric-learning#309]
I have been training a model using with PyTorch Metric Learning and another version of the same model without it, using [this loss](https://github.com/wujiyang/Face_Pytorch/blob/master/margin/ArcMarginProduct.py). And for some reason, the final performance of my PyTorch Metric Learning was being quite worse. I then implemented the same loss as follows:

```
class MarginLoss(WeightRegularizerMixin, BaseMetricLossFunction, metaclass=abc.ABCMeta):
    def __init__(
        self,
        num_classes: int,
        embedding_size: int,
        **kwargs
    ):
        super().__init__(**kwargs)

        self.in_features = embedding_size
        self.out_features = num_classes

        self.weight = nn.Parameter(torch.FloatTensor(self.out_features, self.in_features))
        self.weight_init_func(self.weight)

        self.cross_entropy = nn.CrossEntropyLoss(reduction="none")

    def get_default_weight_init_func(self):
        return c_f.TorchInitWrapper(torch.nn.init.xavier_uniform_)

    def cast_types(self, dtype, device):
        self.weight.data = c_f.to_device(self.weight.data, device=device, dtype=dtype)

    @abc.abstractmethod
    def calculate_logits(self, embeddings: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
        pass

    def compute_loss(self, embeddings: torch.Tensor, labels: torch.Tensor, indices_tuple) -> dict:
        dtype, device = embeddings.dtype, embeddings.device
        self.cast_types(dtype, device)
        miner_weights = lmu.convert_to_weights(indices_tuple, labels, dtype=embeddings.dtype)

        unweighted_loss = self.cross_entropy(self.calculate_logits(embeddings, labels), labels)
        miner_weighted_loss = unweighted_loss * miner_weights

        loss_dict = {
            "loss": {
                "losses": miner_weighted_loss,
                "indices": c_f.torch_arange_from_size(embeddings),
                "reduction_type": "element",
            }
        }
        self.add_weight_regularization_to_loss_dict(loss_dict, self.weight.t())
        return loss_dict

class ArcMarginProductLoss(MarginLoss):
    def __init__(
        self,
        num_classes: int,
        embedding_size: int,
        s: float = 30.0,
        m: float = 0.50,
        easy_margin: bool = False,
        ls_eps: float = 0.0,
        **kwargs
    ):
        super().__init__(num_classes, embedding_size, **kwargs)

        self.s = s
        self.m = m
        self.ls_eps = ls_eps  # label smoothing

        self.easy_margin = easy_margin
        self.cos_m = math.cos(m)
        self.sin_m = math.sin(m)
        self.th = math.cos(math.pi - m)
        self.mm = math.sin(math.pi - m) * m

    def calculate_logits(self, embeddings: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
        cosine = F.linear(F.normalize(embeddings), F.normalize(self.weight))
        sine = torch.sqrt(1.0 - torch.pow(cosine, 2))
        phi = cosine * self.cos_m - sine * self.sin_m
        phi = phi.to(dtype=cosine.dtype)
        if self.easy_margin:
            phi = torch.where(cosine > 0, phi, cosine)
        else:
            phi = torch.where(cosine > self.th, phi, cosine - self.mm)
        # --------------------------- convert label to one-hot ---------------------------
        # one_hot = torch.zeros(cosine.size(), requires_grad=True, device='cuda')
        one_hot = torch.zeros(cosine.size(), device=embeddings.device)
        one_hot.scatter_(1, labels.view(-1, 1).long(), 1)
        if self.ls_eps > 0:
            one_hot = (1 - self.ls_eps) * one_hot + self.ls_eps / self.out_features
        # -------------torch.where(out_i = {x_i if condition_i else y_i) -------------
        logits = (one_hot * phi) + ((1.0 - one_hot) * cosine)
        logits *= self.s
        return logits
```

And even after using the same loss, my results with PyTorch Metric Learning was much worse. After debugging, I've noticed a single difference: the weights of the final layer (that goes through the CrossEntropyLoss) of my PyTorch Metric Learning version was float16 and the other was float32 (I'm using mixed precision to train both).

Then, I remembered that I've read somewhere that the final layer that goes before a Softmax shouldn't output float16. After searching, I've found this:

"Adding a float16 softmax in the middle of a model is fine, but a softmax at the end of the model should be in float32. The reason is that if the intermediate tensor flowing from the softmax to the loss is float16 or bfloat16, numeric issues may occur." from https://www.tensorflow.org/guide/mixed_precision

It is from TensorFlow docs, but I guess it also applies to PyTorch. So, I'm going to stop it from converting the W to float16 and do some tests and get back here to report.

**Comment by KevinMusgrave** [comment]
What version of pytorch-metric-learning are you using? Can you try the latest pre-release and see if the issue persists?

```python
pip install pytorch-metric-learning==0.9.99.dev0
```

**Comment by fernandocamargoti** [comment]
Hello, @KevinMusgrave. I've debugger a little more and discovered that my workaround to [this issue](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/300), which was turning off the `autocast` is responsible for the bug. When the autocast is enabled, the `c_f.to_device` doesn't change the tensors from float16 to float32.

I'll try this version you said anyway to solve that issue.

**Comment by KevinMusgrave** [comment]
If you get around to trying the pre-release, let me know if it works

**Comment by fernandocamargoti** [comment]
It does. Thank you, @KevinMusgrave.
